File: sdk/runanywhere-flutter/lib/foundation/error_types/sdk_error.dart
/// Main SDK error type
/// Matches iOS RunAnywhereError from Foundation/Errors/RunAnywhereError.swift
///
/// Note: Also exported as [RunAnywhereError] for iOS parity
class SDKError implements Exception {
  final String message;
  final SDKErrorType type;

  /// The underlying error that caused this SDK error (if any)
  /// Matches iOS RunAnywhereError.underlyingError
  final Object? underlyingError;

  SDKError(this.message, this.type, {this.underlyingError});

  @override
  String toString() => 'SDKError($type): $message';

  /// Recovery suggestion for the error
  /// Matches iOS RunAnywhereError.recoverySuggestion
  String? get recoverySuggestion {
    switch (type) {
      case SDKErrorType.notInitialized:
        return 'Call RunAnywhere.initialize() before using the SDK.';
      case SDKErrorType.alreadyInitialized:
        return 'The SDK is already initialized. You can use it directly.';
      case SDKErrorType.invalidAPIKey:
        return 'Provide a valid API key in the configuration.';
      case SDKErrorType.invalidConfiguration:
        return 'Check your configuration settings and ensure all required fields are provided.';
      case SDKErrorType.environmentMismatch:
        return 'Use .development or .staging for DEBUG builds. Production environment requires a Release build.';

      case SDKErrorType.modelNotFound:
        return 'Check the model identifier or download the model first.';
      case SDKErrorType.modelLoadFailed:
        return 'Ensure the model file is not corrupted and is compatible with your device.';
      case SDKErrorType.loadingFailed:
        return 'The loading operation failed. Check logs for details.';
      case SDKErrorType.modelValidationFailed:
        return 'The model file may be corrupted or incompatible. Try re-downloading.';
      case SDKErrorType.modelIncompatible:
        return 'Use a different model that is compatible with your device.';
      case SDKErrorType.frameworkNotAvailable:
        return 'Use a different model or device that supports this feature.';

      case SDKErrorType.generationFailed:
        return 'Check your input and try again.';
      case SDKErrorType.generationTimeout:
        return 'Try with a shorter prompt or fewer tokens.';
      case SDKErrorType.contextTooLong:
        return 'Reduce the context size or use a model with larger context window.';
      case SDKErrorType.tokenLimitExceeded:
        return 'Reduce the number of tokens requested.';
      case SDKErrorType.costLimitExceeded:
        return 'Increase your cost limit or use a more cost-effective model.';

      case SDKErrorType.networkError:
      case SDKErrorType.networkUnavailable:
      case SDKErrorType.requestFailed:
      case SDKErrorType.serverError:
        return 'Check your internet connection and try again.';
      case SDKErrorType.downloadFailed:
        return 'Check your internet connection and available storage space.';
      case SDKErrorType.timeout:
        return 'The operation timed out. Try again or check your network connection.';

      case SDKErrorType.storageError:
      case SDKErrorType.insufficientStorage:
        return 'Free up storage space on your device.';
      case SDKErrorType.storageFull:
        return 'Delete unnecessary files to free up space.';

      case SDKErrorType.hardwareUnsupported:
        return 'Use a different model or device that supports this feature.';
      case SDKErrorType.memoryPressure:
        return 'Close other apps to free up memory.';
      case SDKErrorType.thermalStateExceeded:
        return 'Wait for the device to cool down before trying again.';

      case SDKErrorType.componentNotReady:
      case SDKErrorType.componentNotInitialized:
        return 'Ensure the component is properly initialized before use.';
      case SDKErrorType.invalidState:
        return 'Check the current state and ensure operations are called in the correct order.';

      case SDKErrorType.authenticationFailed:
        return 'Check your credentials and try again.';

      case SDKErrorType.databaseInitializationFailed:
        return 'Try reinstalling the app or clearing app data.';

      case SDKErrorType.validationFailed:
        return 'Check your input parameters and ensure they are valid.';
      case SDKErrorType.unsupportedModality:
        return 'Check your input parameters and ensure they are valid.';

      case SDKErrorType.featureNotAvailable:
      case SDKErrorType.notImplemented:
        return 'This feature may be available in a future update.';
    }
  }

  // Factory constructors for common errors
  static SDKError notInitialized([String? message]) {
    return SDKError(
      message ?? 'SDK has not been initialized',
      SDKErrorType.notInitialized,
    );
  }

  static SDKError alreadyInitialized([String? message]) {
    return SDKError(
      message ?? 'SDK is already initialized',
      SDKErrorType.alreadyInitialized,
    );
  }

  static SDKError invalidAPIKey([String? message]) {
    return SDKError(
      message ?? 'Invalid API key',
      SDKErrorType.invalidAPIKey,
    );
  }

  static SDKError invalidConfiguration([String? message]) {
    return SDKError(
      message ?? 'Invalid configuration',
      SDKErrorType.invalidConfiguration,
    );
  }

  static SDKError environmentMismatch([String? message]) {
    return SDKError(
      message ?? 'Environment configuration mismatch',
      SDKErrorType.environmentMismatch,
    );
  }

  static SDKError modelNotFound(String modelId) {
    return SDKError(
      'Model not found: $modelId',
      SDKErrorType.modelNotFound,
    );
  }

  static SDKError modelLoadFailed(String modelId, Object error) {
    return SDKError(
      'Model load failed: $modelId - $error',
      SDKErrorType.modelLoadFailed,
      underlyingError: error,
    );
  }

  static SDKError loadingFailed([String? message]) {
    return SDKError(
      message ?? 'Loading failed',
      SDKErrorType.loadingFailed,
    );
  }

  static SDKError featureNotAvailable([String? message]) {
    return SDKError(
      message ?? 'Feature not available',
      SDKErrorType.featureNotAvailable,
    );
  }

  static SDKError componentNotReady(String component) {
    return SDKError(
      'Component not ready: $component',
      SDKErrorType.componentNotReady,
    );
  }

  static SDKError networkError([String? message]) {
    return SDKError(
      message ?? 'Network error occurred',
      SDKErrorType.networkError,
    );
  }

  static SDKError storageError([String? message]) {
    return SDKError(
      message ?? 'Storage error occurred',
      SDKErrorType.storageError,
    );
  }

  static SDKError validationFailed([String? message]) {
    return SDKError(
      message ?? 'Validation failed',
      SDKErrorType.validationFailed,
    );
  }

  static SDKError invalidState([String? message]) {
    return SDKError(
      message ?? 'Invalid state',
      SDKErrorType.invalidState,
    );
  }

  static SDKError timeout([String? message]) {
    return SDKError(
      message ?? 'Operation timed out',
      SDKErrorType.timeout,
    );
  }

  static SDKError downloadFailed([String? message]) {
    return SDKError(
      message ?? 'Download failed',
      SDKErrorType.downloadFailed,
    );
  }

  static SDKError generationFailed([String? message]) {
    return SDKError(
      message ?? 'Generation failed',
      SDKErrorType.generationFailed,
    );
  }

  static SDKError generationTimeout([String? message]) {
    return SDKError(
      message ?? 'Generation timed out',
      SDKErrorType.generationTimeout,
    );
  }

  static SDKError contextTooLong(int provided, int maximum) {
    return SDKError(
      'Context too long: $provided tokens (maximum: $maximum)',
      SDKErrorType.contextTooLong,
    );
  }

  static SDKError tokenLimitExceeded(int requested, int maximum) {
    return SDKError(
      'Token limit exceeded: requested $requested, maximum $maximum',
      SDKErrorType.tokenLimitExceeded,
    );
  }

  static SDKError costLimitExceeded(double estimated, double limit) {
    return SDKError(
      'Cost limit exceeded: estimated \$${estimated.toStringAsFixed(2)}, limit \$${limit.toStringAsFixed(2)}',
      SDKErrorType.costLimitExceeded,
    );
  }

  static SDKError authenticationFailed([String? message]) {
    return SDKError(
      message ?? 'Authentication failed',
      SDKErrorType.authenticationFailed,
    );
  }

  static SDKError databaseInitializationFailed(Object error) {
    return SDKError(
      'Database initialization failed: $error',
      SDKErrorType.databaseInitializationFailed,
      underlyingError: error,
    );
  }

  static SDKError unsupportedModality([String? message]) {
    return SDKError(
      message ?? 'Unsupported modality',
      SDKErrorType.unsupportedModality,
    );
  }

  static SDKError insufficientStorage(int required, int available) {
    return SDKError(
      'Insufficient storage: $required bytes required, $available bytes available',
      SDKErrorType.insufficientStorage,
    );
  }

  static SDKError hardwareUnsupported([String? message]) {
    return SDKError(
      message ?? 'Hardware not supported',
      SDKErrorType.hardwareUnsupported,
    );
  }

  static SDKError serverError([String? message]) {
    return SDKError(
      message ?? 'Server error occurred',
      SDKErrorType.serverError,
    );
  }

  static SDKError requestFailed(Object error) {
    return SDKError(
      'Request failed: $error',
      SDKErrorType.requestFailed,
      underlyingError: error,
    );
  }

  static SDKError networkUnavailable([String? message]) {
    return SDKError(
      message ?? 'Network connection unavailable',
      SDKErrorType.networkUnavailable,
    );
  }

  static SDKError modelValidationFailed(String modelId, List<String> errors) {
    return SDKError(
      'Model \'$modelId\' validation failed: ${errors.join(', ')}',
      SDKErrorType.modelValidationFailed,
    );
  }

  static SDKError modelIncompatible(String modelId, String reason) {
    return SDKError(
      'Model \'$modelId\' is incompatible: $reason',
      SDKErrorType.modelIncompatible,
    );
  }

  static SDKError frameworkNotAvailable([String? message]) {
    return SDKError(
      message ?? 'Required framework not available',
      SDKErrorType.frameworkNotAvailable,
    );
  }
}

/// SDK error types
/// Matches iOS RunAnywhereError cases
enum SDKErrorType {
  // Initialization errors
  notInitialized,
  alreadyInitialized,
  invalidAPIKey,
  invalidConfiguration,
  environmentMismatch,

  // Model errors
  modelNotFound,
  modelLoadFailed,
  loadingFailed, // Separate from modelLoadFailed for iOS parity
  modelValidationFailed,
  modelIncompatible,
  frameworkNotAvailable,

  // Generation errors
  generationFailed,
  generationTimeout,
  contextTooLong,
  tokenLimitExceeded,
  costLimitExceeded,

  // Network errors
  networkError,
  networkUnavailable,
  requestFailed,
  downloadFailed,

  // Storage errors
  storageError,
  insufficientStorage,
  storageFull,

  // Hardware errors
  hardwareUnsupported,
  memoryPressure,
  thermalStateExceeded,

  // Component errors
  componentNotReady,
  componentNotInitialized,

  // Authentication errors
  authenticationFailed,

  // Database errors
  databaseInitializationFailed,

  // Feature errors
  featureNotAvailable,
  notImplemented,

  // Validation errors
  validationFailed,
  unsupportedModality,

  // General errors
  invalidState,
  timeout,
  serverError,
}

/// Type alias for iOS parity
/// iOS uses RunAnywhereError; this alias provides compatibility
typedef RunAnywhereError = SDKError;

/// Type alias for iOS parity
/// iOS uses ErrorCategory; this alias provides compatibility
typedef ErrorCategory = SDKErrorType;

File: sdk/runanywhere-react-native/src/errors/SDKError.ts
/**
 * RunAnywhere React Native SDK - Error Types
 *
 * These error types match the iOS Swift SDK error definitions.
 * Reference: sdk/runanywhere-swift/Sources/RunAnywhere/Public/Errors/SDKError.swift
 */

import { LLMFramework } from '../types';

/**
 * Error codes matching the iOS SDK error cases
 */
export enum SDKErrorCode {
  NotInitialized = 'NOT_INITIALIZED',
  NotImplemented = 'NOT_IMPLEMENTED',
  InvalidAPIKey = 'INVALID_API_KEY',
  ModelNotFound = 'MODEL_NOT_FOUND',
  ModelNotLoaded = 'MODEL_NOT_LOADED',
  ModelLoadFailed = 'MODEL_LOAD_FAILED',
  LoadingFailed = 'LOADING_FAILED',
  GenerationFailed = 'GENERATION_FAILED',
  GenerationTimeout = 'GENERATION_TIMEOUT',
  FrameworkNotAvailable = 'FRAMEWORK_NOT_AVAILABLE',
  DownloadFailed = 'DOWNLOAD_FAILED',
  ValidationFailed = 'VALIDATION_FAILED',
  RoutingFailed = 'ROUTING_FAILED',
  DatabaseInitializationFailed = 'DATABASE_INITIALIZATION_FAILED',
  UnsupportedModality = 'UNSUPPORTED_MODALITY',
  InvalidResponse = 'INVALID_RESPONSE',
  AuthenticationFailed = 'AUTHENTICATION_FAILED',
  NetworkError = 'NETWORK_ERROR',
  InvalidState = 'INVALID_STATE',
  ComponentNotInitialized = 'COMPONENT_NOT_INITIALIZED',
  ComponentNotReady = 'COMPONENT_NOT_READY',
  Timeout = 'TIMEOUT',
  ServerError = 'SERVER_ERROR',
  StorageError = 'STORAGE_ERROR',
  InvalidConfiguration = 'INVALID_CONFIGURATION',
  // STT specific errors
  TranscriptionFailed = 'TRANSCRIPTION_FAILED',
  StreamCreationFailed = 'STREAM_CREATION_FAILED',
  ServiceNotInitialized = 'SERVICE_NOT_INITIALIZED',
  // TTS specific errors
  SynthesisFailed = 'SYNTHESIS_FAILED',
  VoiceNotFound = 'VOICE_NOT_FOUND',
  // VAD specific errors
  ProcessingFailed = 'PROCESSING_FAILED',
  CleanupFailed = 'CLEANUP_FAILED',
  Unknown = 'UNKNOWN',
}

/**
 * SDK Error class matching the iOS Swift SDK SDKError enum
 */
export class SDKError extends Error {
  /** Error code for programmatic handling */
  public readonly code: SDKErrorCode;

  /** Additional context or reason */
  public readonly reason?: string;

  /** Underlying error if any */
  public readonly underlyingError?: Error;

  /** Framework involved (for framework-specific errors) */
  public readonly framework?: LLMFramework;

  /** Model ID involved (for model-specific errors) */
  public readonly modelId?: string;

  constructor(
    code: SDKErrorCode,
    message: string,
    options?: {
      reason?: string;
      underlyingError?: Error;
      framework?: LLMFramework;
      modelId?: string;
    }
  ) {
    super(message);
    this.name = 'SDKError';
    this.code = code;
    this.reason = options?.reason;
    this.underlyingError = options?.underlyingError;
    this.framework = options?.framework;
    this.modelId = options?.modelId;

    // Maintains proper stack trace for where our error was thrown
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, SDKError);
    }
  }

  /**
   * Create a human-readable error description
   */
  get errorDescription(): string {
    return this.message;
  }

  // ============================================================================
  // Static Factory Methods (matching iOS SDK error cases)
  // ============================================================================

  static notInitialized(): SDKError {
    return new SDKError(
      SDKErrorCode.NotInitialized,
      'SDK not initialized. Call initialize() first.'
    );
  }

  static notImplemented(): SDKError {
    return new SDKError(
      SDKErrorCode.NotImplemented,
      'This feature is not yet implemented.'
    );
  }

  static invalidAPIKey(reason: string): SDKError {
    return new SDKError(SDKErrorCode.InvalidAPIKey, `Invalid API key: ${reason}`, {
      reason,
    });
  }

  static modelNotFound(modelId: string): SDKError {
    return new SDKError(SDKErrorCode.ModelNotFound, `Model '${modelId}' not found.`, {
      modelId,
    });
  }

  static loadingFailed(reason: string, modelId?: string): SDKError {
    return new SDKError(SDKErrorCode.LoadingFailed, `Failed to load model: ${reason}`, {
      reason,
      modelId,
    });
  }

  static generationFailed(reason: string): SDKError {
    return new SDKError(SDKErrorCode.GenerationFailed, `Generation failed: ${reason}`, {
      reason,
    });
  }

  static generationTimeout(reason: string): SDKError {
    return new SDKError(
      SDKErrorCode.GenerationTimeout,
      `Generation timed out: ${reason}`,
      { reason }
    );
  }

  static frameworkNotAvailable(framework: LLMFramework): SDKError {
    return new SDKError(
      SDKErrorCode.FrameworkNotAvailable,
      `Framework ${framework} not available`,
      { framework }
    );
  }

  static downloadFailed(error: Error, modelId?: string): SDKError {
    return new SDKError(
      SDKErrorCode.DownloadFailed,
      `Download failed: ${error.message}`,
      { underlyingError: error, modelId }
    );
  }

  static validationFailed(reason: string): SDKError {
    return new SDKError(SDKErrorCode.ValidationFailed, `Validation failed: ${reason}`, {
      reason,
    });
  }

  static routingFailed(reason: string): SDKError {
    return new SDKError(SDKErrorCode.RoutingFailed, `Routing failed: ${reason}`, {
      reason,
    });
  }

  static databaseInitializationFailed(error: Error): SDKError {
    return new SDKError(
      SDKErrorCode.DatabaseInitializationFailed,
      `Database initialization failed: ${error.message}`,
      { underlyingError: error }
    );
  }

  static unsupportedModality(modality: string): SDKError {
    return new SDKError(
      SDKErrorCode.UnsupportedModality,
      `Unsupported modality: ${modality}`,
      { reason: modality }
    );
  }

  static invalidResponse(reason: string): SDKError {
    return new SDKError(SDKErrorCode.InvalidResponse, `Invalid response: ${reason}`, {
      reason,
    });
  }

  static authenticationFailed(reason: string): SDKError {
    return new SDKError(
      SDKErrorCode.AuthenticationFailed,
      `Authentication failed: ${reason}`,
      { reason }
    );
  }

  static networkError(reason: string): SDKError {
    return new SDKError(SDKErrorCode.NetworkError, `Network error: ${reason}`, {
      reason,
    });
  }

  static invalidState(reason: string): SDKError {
    return new SDKError(SDKErrorCode.InvalidState, `Invalid state: ${reason}`, {
      reason,
    });
  }

  static componentNotInitialized(component: string): SDKError {
    return new SDKError(
      SDKErrorCode.ComponentNotInitialized,
      `Component not initialized: ${component}`,
      { reason: component }
    );
  }

  static timeout(reason: string): SDKError {
    return new SDKError(SDKErrorCode.Timeout, `Operation timed out: ${reason}`, {
      reason,
    });
  }

  static serverError(reason: string): SDKError {
    return new SDKError(SDKErrorCode.ServerError, `Server error: ${reason}`, {
      reason,
    });
  }

  static storageError(reason: string): SDKError {
    return new SDKError(SDKErrorCode.StorageError, `Storage error: ${reason}`, {
      reason,
    });
  }

  static invalidConfiguration(reason: string): SDKError {
    return new SDKError(
      SDKErrorCode.InvalidConfiguration,
      `Invalid configuration: ${reason}`,
      { reason }
    );
  }

  /**
   * Create an SDKError from a native error payload
   */
  static fromNative(nativeError: {
    code?: string;
    message?: string;
    reason?: string;
    framework?: string;
    modelId?: string;
  }): SDKError {
    const code =
      (nativeError.code as SDKErrorCode) || SDKErrorCode.Unknown;
    const message = nativeError.message || 'An unknown error occurred';

    return new SDKError(code, message, {
      reason: nativeError.reason,
      framework: nativeError.framework as LLMFramework | undefined,
      modelId: nativeError.modelId,
    });
  }

  /**
   * Convert to a plain object for serialization
   */
  toJSON(): Record<string, unknown> {
    return {
      name: this.name,
      code: this.code,
      message: this.message,
      reason: this.reason,
      framework: this.framework,
      modelId: this.modelId,
    };
  }
}

// Re-export error code for convenience
export { SDKErrorCode as ErrorCode };
